### PR TITLE
Add db_backup cronjob

### PR DESCRIPTION
**Pull request by CypherpunkSamurai**
# Description

Added db_backup cronjob to backup the db to local file storage.

## Related issues
#16

## Type of change

- [X] New feature (non-breaking change which adds functionality).

# Checklist

- [X] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [X] The pull request is for the branch `develop`
- [ ] The tests gave 0 errors.
- [X] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] The commits were squashed into a single one (optional, they will be squashed anyway by the maintainer)
  
### Important Rules
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review

### CODE REVIEW

File: greedybear/settings.py [hunk 1]
```diff
@@ -13,7 +13,14 @@
 SECRET_KEY = os.environ.get("DJANGO_SECRET", None) or get_random_secret_key()
 
 # SECURITY WARNING: don't run with debug turned on in production!
-DEBUG = os.environ.get("DEBUG", False) == "True"
+# DEBUG = os.environ.get("DEBUG", False) == "True"
+DEBUG = True
```
Comment (uzaxirr) [inline greedybear/settings.py]: Kindly revert this change

File: greedybear/settings.py [hunk 2]
```diff
@@ -13,7 +13,14 @@
...
+
+# DB Backup Configs
+DBBACKUP_STORAGE = "django.core.files.storage.FileSystemStorage"
+DBBACKUP_STORAGE_OPTIONS = {"location": "/var/log/greedybear/"}
+# Database Backup Interval (in minutes). default 120 mins = 2 hours.
+DBBACKUP_FREQUENCY = int(os.environ.get("DBBACKUP_FREQUENCY", 120)) == 120
```
Comment (uzaxirr) [inline greedybear/settings.py]: Not sure if it matters much but instead of `DBBACKUP_STORAGE` you can use `DB_BACKUP_STORAGE` and similarly in other variables as well this looks much readable

**Review by uzaxirr — CHANGES_REQUESTED** [review]

File: greedybear/settings.py (hunk 2 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Ah, yes. I was honestly thinking about it while naming it. Sure, will do.

Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Oh, I just remembered. The module depends on those settings [\[ref\]](https://django-dbbackup.readthedocs.io/en/master/installation.html#add-it-in-your-project), so I don't think its possible to rename them (except `DBBACKUP_FREQUENCY`, which I manually added)

Comment (uzaxirr) [reply greedybear/settings.py]: Ohh i see, that's their default name. It's fine then.

File: greedybear/settings.py (hunk 1 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Oops. Really Sorry, forgot about it.

File: greedybear/settings.py (hunk 2 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Should I change `DBBACKUP_FREQUENCY` or should I keep it as is to represent common configs?

Comment (uzaxirr) [reply greedybear/settings.py]: I recommend not to change in order to keep all the DB Backup related vars in a uniform manner.

Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Ok Sir, Got it 👍🏼

File: greedybear/settings.py [hunk 3]
```diff
@@ -15,6 +15,12 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = os.environ.get("DEBUG", False) == "True"
 
+# DB Backup Configs
+DBBACKUP_STORAGE = "django.core.files.storage.FileSystemStorage"
+DBBACKUP_STORAGE_OPTIONS = {"location": "/var/log/greedybear/"}
+# Database Backup Interval (in minutes). default 120 mins = 2 hours.
+DBBACKUP_FREQUENCY = int(os.environ.get("DBBACKUP_FREQUENCY", 120)) or 120
```
Comment (mlodic) [inline greedybear/settings.py]: imo once a day it is enough

File: greedybear/settings.py [hunk 4]
```diff
@@ -15,6 +15,12 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = os.environ.get("DEBUG", False) == "True"
 
+# DB Backup Configs
+DBBACKUP_STORAGE = "django.core.files.storage.FileSystemStorage"
+DBBACKUP_STORAGE_OPTIONS = {"location": "/var/log/greedybear/"}
```
Comment (mlodic) [inline greedybear/settings.py]: these are not logs, I would leverage a new directory like `/opt/backups`

File: greedybear/settings.py [hunk 5]
```diff
@@ -56,6 +62,8 @@
     "gui.apps.GuiConfig",
     "api.apps.ApiConfig",
     "greedybear.apps.GreedyBearConfig",
+    # db backups
+    "dbbackup",
```
Comment (mlodic) [inline greedybear/settings.py]: can you explain advantages of this external app when we have https://docs.djangoproject.com/en/4.0/ref/django-admin/#dumpdata as default in Django?

File: greedybear/cronjobs/db_backup.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+from django.core import management
+
+from greedybear import settings
+from greedybear.cronjobs.base import Cronjob
+
+
+class DbBackup(Cronjob):
```
Comment (mlodic) [inline greedybear/cronjobs/db_backup.py:12]: without adding this cron as a celery task, it would never run

File: greedybear/settings.py (hunk 4 above)
Comment (mlodic) [reply greedybear/settings.py]: we also need to provide a mechanism for deletion of too old files in this directory.

I would say a maximum of 10 backups we keep there as default and then we delete the old ones

**Review by mlodic — CHANGES_REQUESTED** [review]

File: greedybear/settings.py (hunk 5 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Oh, I haven't looked into that yet. That seems like a better alternative than adding another dependency. I'll try implementing it with `dumpdata` instead 👍🏼

File: greedybear/cronjobs/db_backup.py (hunk 6 above)
Comment (CypherpunkSamurai) [reply greedybear/cronjobs/db_backup.py:12]: Ok Sir.

File: greedybear/settings.py (hunk 4 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: I've fixed that with the recent commit. Please check

File: greedybear/settings.py (hunk 5 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Hello Sir, I've updated the implementation with `dumpdata`. Please review changes :)

**Comment by CypherpunkSamurai** [comment]
Hello Sir, I've re-implemented it with `dumpdata` please check :)

File: greedybear/settings.py [hunk 7]
```diff
@@ -15,6 +15,12 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = os.environ.get("DEBUG", False) == "True"
 
+# DB Backup Config
+# Backup Interval in minutes. Default: 1440 min = 1 day.
+DB_BACKUP_FREQUENCY = int(os.environ.get("DB_BACKUP_FREQUENCY", 1440)) or 1440
+# Backup Storage Location and Max Backups. (When Max Backup is reached old backups are removed).
+DB_BACKUP_STORAGE_OPTIONS = {"location": "/opt/backups", "max_backups": 10}
```
Comment (mlodic) [inline greedybear/settings.py]: I think we can have 2 different variables for those, each configurable like the previous one.

File: greedybear/settings.py [hunk 8]
```diff
@@ -15,6 +15,12 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = os.environ.get("DEBUG", False) == "True"
 
+# DB Backup Config
+# Backup Interval in minutes. Default: 1440 min = 1 day.
+DB_BACKUP_FREQUENCY = int(os.environ.get("DB_BACKUP_FREQUENCY", 1440)) or 1440
```
Comment (mlodic) [inline greedybear/settings.py]: have you tried to customize this value? cause this does not work.

you have to leverage the env file that is loaded by the application

File: greedybear/cronjobs/db_backup.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+    def _remove_oldest_backup(self, dir):
+        # removes files older than self.max_backups
+        files = os.listdir(dir)
+        if len(files) >= self.max_backups:
+            oldest_file = min(files, key=os.path.getctime)
+            os.remove(os.path.abspath(oldest_file))
+
+    def run(self):
+        self._remove_oldest_backup(self.location)
```
Comment (mlodic) [inline greedybear/cronjobs/db_backup.py]: no need to pass this value as a variable...

File: greedybear/cronjobs/db_backup.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            str(settings.DB_BACKUP_STORAGE_OPTIONS["location"]) or "/opt/backups"
+        )
+        self.max_backups = int(settings.DB_BACKUP_STORAGE_OPTIONS["max_backups"]) or 10
+
+    @property
+    def minutes_back_to_lookup(self):
+        return self.interval
+
+    def _remove_oldest_backup(self, dir):
+        # removes files older than self.max_backups
```
Comment (mlodic) [inline greedybear/cronjobs/db_backup.py:24]: add log

File: greedybear/cronjobs/db_backup.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+            f"{self.location}/db_backup_{datetime.now().strftime('%Y%m%d-%H%M%S')}.json"
+        )
+        # Init a new buffer
+        buf = StringIO()
+        # Call command with and pipe stdout to buffer
+        management.call_command("dbbackup", app_name, stdout=buf)
+        buf.seek(0)
+        # Write buffer to file
+        with open(backup_file, "w", "utf-8") as dump:
+            dump.write(buf.read())
```
Comment (mlodic) [inline greedybear/cronjobs/db_backup.py:47]: can you show an example of a successful run of this cron?

plus, there are no logs in all the cron, it is impossible to debug what is happening without them

File: greedybear/celery.py [hunk 12]
```diff
@@ -86,4 +86,10 @@ def setup_loggers(*args, **kwargs):
         "schedule": crontab(minute=33),
         "options": {"queue": "default"},
     },
+    # once a day
```
Comment (mlodic) [inline greedybear/celery.py]: ok but the DB_BACKUP_FREQUENCY varialbe is never used. It will always be executed once a day...

**Review by mlodic — CHANGES_REQUESTED** [review]

File: greedybear/settings.py (hunk 7 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Ok Sir.

File: greedybear/cronjobs/db_backup.py (hunk 9 above)
Comment (CypherpunkSamurai) [reply greedybear/cronjobs/db_backup.py]: Oh Right we can access it from the method itself 👍🏼

File: greedybear/cronjobs/db_backup.py (hunk 10 above)
Comment (CypherpunkSamurai) [reply greedybear/cronjobs/db_backup.py:24]: Ok Sir.

File: greedybear/settings.py (hunk 8 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Ok Sir. I'll try debugging it.

File: greedybear/celery.py (hunk 12 above)
Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Sir, I've added the backup frequency to the celery.py. It should work now. Please review.

File: greedybear/celery.py [hunk 13]
```diff
@@ -86,4 +88,10 @@ def setup_loggers(*args, **kwargs):
...
+    "db_backup": {
+        "task": "greedybear.tasks.db_backup",
+        "schedule": crontab(minute=DB_BACKUP_FREQUENCY),
```
Comment (mlodic) [inline greedybear/celery.py]: this does not work as you intend. Same syntax as a normal cronjob is required

File: greedybear/settings.py [hunk 14]
```diff
@@ -15,6 +15,13 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = os.environ.get("DEBUG", False) == "True"
 
+# DB Backup Config
+# Backup Interval in minutes. Default: 1440 min = 1 day.
+DB_BACKUP_FREQUENCY = int(os.getenv("DB_BACKUP_FREQUENCY", 1440))
```
Comment (mlodic) [inline greedybear/settings.py]: it is enough to allow users to just choose to do the download every X days. Then we set static the hour/minute time

File: greedybear/celery.py [hunk 15]
```diff
@@ -10,6 +10,8 @@
 from django.conf import settings
 from kombu import Exchange, Queue
 
+from greedybear.settings import DB_BACKUP_FREQUENCY
```
Comment (mlodic) [inline greedybear/celery.py]: no need, we already import `from django.conf import settings`

**Review by mlodic — CHANGES_REQUESTED** [review]

File: greedybear/celery.py (hunk 15 above)
Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Ok Sir

File: greedybear/settings.py (hunk 14 above)
Comment (CypherpunkSamurai) [reply greedybear/settings.py]: Ok Sir. I'll add that

File: greedybear/celery.py (hunk 13 above)
Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Ok Sir, I've updated it with a manual crontab format. Using: `f"0 0 * * */{settings.DB_BACKUP_FREQUENCY}"`.

Comment (mlodic) [reply greedybear/celery.py]: please check the documentation: https://docs.celeryq.dev/en/stable/userguide/periodic-tasks.html?highlight=cron

Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Ok, I've updated it now according to the documentation sir. Please check.

File: greedybear/celery.py [hunk 16]
```diff
@@ -86,4 +86,10 @@ def setup_loggers(*args, **kwargs):
         "schedule": crontab(minute=33),
         "options": {"queue": "default"},
     },
+    # run once every settings.DB_BACKUP_FREQUENCY * 24 hour
+    "db_backup": {
+        "task": "greedybear.tasks.db_backup",
+        "schedule": crontab(minute=0, hour=f'*/{24 * settings.DB_BACKUP_FREQUENCY}'),
```
Comment (mlodic) [inline greedybear/celery.py]: this is still wrong, hour and minute must be fixes, days should change

Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Sir I currently had a doubt about the `days` parameter. Sir If you don't mind me asking, isn't the days parameter for 'day of the month' or 'day of the week'? I looked into a few cronjob syntaxes and it all suggested using either `hours` or `days=*/DAYS_INTERVAL`

Comment (mlodic) [reply greedybear/celery.py]: maybe we could just leave this to do it once every day

Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Ok Sir.

I'm very new to celery and cron jobs, I'll try my best to learn it as soon as possible to be capable of working on Greedybear.

Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Hello Sir, I just noticed that celery schedules support `datetime.timedelta` formats. I replaced the `crontab()` implementation with `timedelta(days=....)` implementation. Is this ok, or should we resort to settings daily cron jobs?

Comment (mlodic) [reply greedybear/celery.py]: where did you get this? have you tried it working?

Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Sir I got that from the celery documentation Available Feilds section. I haven't yet tested it, I'll update soon once I test it.

https://docs.celeryq.dev/en/stable/userguide/periodic-tasks.html?highlight=cron#available-fields

Comment (CypherpunkSamurai) [reply greedybear/celery.py]: I tried to test the cron job, but there seems to some problem. Maybe I'm running it incorrectly.

I'm using the following commands for testing:
```
export DEBUG=True
docker-compose -p greedybear build
docker-compose -p greedybear up
```

I'm not sure why but my docker runs dont print any logs from celery cron jobs. I've even tried adding `print()` to current cron job, and other cron jobs, but none of them seem to be running.

Comment (mlodic) [reply greedybear/celery.py]: why don't you just populate the `env_file` inside the `docker` folder with that DEBUG=True value?

anyway the logs should work and should appear in the `greedybear.log` file

Comment (mlodic) [reply greedybear/celery.py]: can you please explain which are your tests? cause you have implemented no tests here

Comment (CypherpunkSamurai) [reply greedybear/celery.py]: I've edited the env file this time and tried looking up the log files. the celery logs aren't present there.

> can you please explain which are your tests? cause you have implemented no tests here

Oh, I was actually referring to changing the celery schedules for all the cron jobs during debug. Just for testing if the cron jobs are working at all. It turns out the cron jobs are not working for some reason, or i'm looking for logs in the wrong place.

Comment (mlodic) [reply greedybear/celery.py]: It is difficult to guess from here why in your env celery does not work while in mine yes without any problems...in every case, there are no tests implemented for this thing. We need them so we can accept the PR. See in the `tests` folder and try to prove that what you did work, without thinking about celery

Comment (CypherpunkSamurai) [reply greedybear/celery.py]: Ok Sir. I'll try it. Also, are there any other ways to check if the celery cron job is running? or maybe manually trigger it?

Comment (mlodic) [reply greedybear/celery.py]: this are just some basic hints:

- You can set up a time that is close to yours if you want to see that the cron works
- you can use `.delay()` from `celery` to have a task called via celery. This excludes that there is a problem in celery but does not exclude that it could be one from the cronjob configuration
- a proper test will exclude bugs in normal launches, even without celery. So this is primary importance
